libgdf/src/type_dispatcher.cuh

```diff
@@ -0,0 +1,126 @@
+ *
+ * A strong typedef  provides a new, concrete type, unlike a normal C++ typedef which
+ * is simply a type alias. These "strong typedef" structs simply wrap a single member
+ * variable of a fundamental type called 'value'. In order to access the underlying 
+ * value, one must use the "unwrap" function which will provide a reference to the 
+ * underlying value.
+ *
+ * See types.cuh for more detail.
+ *
+ * Example usage with a functor that returns the size of the dispatched type:
+ *
+ * struct example_functor{
+ *  template <typename col_type>
+ *  int operator()(){
+ *    return sizeof(T);
+ *  }
+ * };
+ *
+ * gdf::type_dispatcher(GDF_INT8, example_functor);  // returns 1
+ * gdf::type_dispatcher(GDF_INT64, example_functor); // returns 8
+ *
+ * Example usage of of the "unwrap" function in a functor for checking if element "i" 
+ * in column "lhs" is equal to element "j" in column "rhs":
+ *
+ * struct elements_are_equal{
+ *   template <typename col_type>
+ *   bool operator()(void const * lhs, int i,
+ *                   void const * rhs, int j)
+ *   {
+ *     // Cast the void* data buffer to the dispatched type and retrieve elements 
+ *     // "i" and "j" from the respective columns
+ *     col_type const i_elem = static_cast<col_type const*>(lhs)[i];
+ *     col_type const j_elem = static_cast<col_type const*>(rhs)[j];
+ *
+ *     // "col_type" may be a wrapper struct. Therefore, use the "unwrap" function
+ *     // to retrieve a reference to the underlying value. If "col_type" is a 
+ *     // fundamental type, "unwrap" simply passes through the same value and
+ *     // is effectively a no-op
+ *     return gdf::detail::unwrap(i_elem) == gdf::detail::unwrap(j_elem);
+ *   }
+ * };
+ *
+ * The return type for all template instantiations of the functor's "operator()" 
+ * lambda must be the same, otherwise there will be a compiler error as you would be
+ * trying to return different types from the same function.
+ *
+ * @Param dtype The gdf_dtype enum that determines which type will be dispatched
+ * @Param f The functor with a templated "operator()" that will be invoked with 
+ * the dispatched type
+ * @Param args A parameter-pack (i.e., arbitrary number of arguments) that will 
+ * be perfectly-forwarded as the arguments of the functor's "operator()".
+ * @tparam enum_map A traits struct templated on a gdf_dtype value with a member 
+ * type called "type" that maps a gdf_dtype to the underlying datatype.
+ *
+ * @Returns Whatever is returned by the functor's "operator()". 
+ *
+ */
+/* ----------------------------------------------------------------------------*/
+// This pragma disables a compiler warning that complains about the valid usage
+// of calling a __host__ functor from this function which is __host__ __device__
+namespace gdf{
+
+#pragma hd_warning_disable
+template < class functor_t, 
+           typename... Ts>
+__host__ __device__ __forceinline__
```

Comment:
I think if you put the __host__ __device__ CUDA specifics in a macro that detects the compiler / compilation mode, you could indeed make this work when included from a .cpp file or compiled with a different compiler, as @eyalroz requested.  Same for types.h?

Reply:
`#ifdef __CUDACC__`, should work for both files (with both NVCC and clang++). And then they could both be `.hpp` files.